File: lib/src/parser/staff_layout_parser.dart
import 'package:musicxml_parser/src/models/staff_layout.dart';
import 'package:musicxml_parser/src/parser/xml_helper.dart';
import 'package:xml/xml.dart';

/// Parser for <staff-layout> elements.
class StaffLayoutParser {
  /// Parses an [XmlElement] into a [StaffLayout] object.
  StaffLayout parse(XmlElement element) {
    // TODO: Implement parsing logic
    // Placeholder implementation
    final numberStr = element.getAttribute('number');
    final staffNumber = numberStr != null ? int.tryParse(numberStr) ?? 1 : 1;

    return StaffLayout(
      staffNumber: staffNumber,
      staffDistance: XmlHelper.getElementTextAsDouble(element.findElements('staff-distance').firstOrNull),
    );
  }
}

File: csharp/MusicXMLParser/Parser/StaffLayoutParser.cs
// Assuming necessary using statements for MusicXML models and helpers
using System.Xml.Linq;
using System.Linq;
using MusicXMLParser.Models; // For StaffLayout
using MusicXMLParser.Utils; // For XmlHelper
using MusicXMLParser.Exceptions; // If needed for validation

namespace MusicXMLParser.Parser
{
    /// <summary>
    /// Parser for <staff-layout> elements.
    /// </summary>
    public class StaffLayoutParser
    {
        public StaffLayout Parse(XElement element)
        {
            var numberStr = element.Attribute("number")?.Value;
            // Default to 1 if attribute is missing or invalid, as per typical MusicXML processor behavior
            // or throw validation exception if strictness is required.
            // The Dart code defaults to 1.
            int staffNumber = 1;
            if (!string.IsNullOrEmpty(numberStr))
            {
                if (int.TryParse(numberStr, out int parsedNumber))
                {
                    staffNumber = parsedNumber;
                }
                // Optionally, add a warning or exception if parsing fails but attribute exists
            }

            var staffDistance = XmlHelper.GetElementTextAsDouble(element.Elements("staff-distance").FirstOrDefault());

            // Assuming StaffLayout constructor can handle nullable staffDistance if it's optional in the model.
            // If staffDistance is required by the model and can't be null,
            // you might need to throw MusicXmlStructureException if it's missing.
            // The Dart model seems to allow null for staffDistance.
            return new StaffLayout(
                staffNumber: staffNumber,
                staffDistance: staffDistance
            );
        }
    }
}
